src/Compilers/CSharp/Portable/Emitter/Model/PEAssemblyBuilder.cs

```diff
@@ -131,6 +150,66 @@ protected override void AddEmbeddedResourcesFromAddedModules(ArrayBuilder<Cci.Ma
         public override string Name => _metadataName;
         public AssemblyIdentity Identity => _sourceAssembly.Identity;
         public Version AssemblyVersionPattern => _sourceAssembly.AssemblyVersionPattern;
+
+        internal override SynthesizedAttributeData SynthesizeEmbeddedAttribute()
+        {
+            if (_embeddedAttributesConstructors.TryGetValue(WellKnownMember.Microsoft_CodeAnalysis_EmbeddedAttribute__ctor, out var constructor))
+            {
+                return new SynthesizedAttributeData(constructor, ImmutableArray<TypedConstant>.Empty, ImmutableArray<KeyValuePair<string, TypedConstant>>.Empty);
+            }
+
+            return base.SynthesizeEmbeddedAttribute();
+        }
+
+        internal override SynthesizedAttributeData SynthesizeIsReadOnlyAttribute()
+        {
+            if (_embeddedAttributesConstructors.TryGetValue(WellKnownMember.System_Runtime_CompilerServices_IsReadOnlyAttribute__ctor, out var constructor))
+            {
+                return new SynthesizedAttributeData(constructor, ImmutableArray<TypedConstant>.Empty, ImmutableArray<KeyValuePair<string, TypedConstant>>.Empty);
+            }
+
+            return base.SynthesizeIsReadOnlyAttribute();
+        }
+
+        private ImmutableDictionary<WellKnownMember, MethodSymbol> CreateEmbeddedAttributesIfNeeded(DiagnosticBag diagnostics)
+        {
+            var builder = ArrayBuilder<KeyValuePair<WellKnownMember, MethodSymbol>>.GetInstance();
+
+            if (_sourceAssembly.DeclaringCompilation.NeedsGeneratedIsReadOnlyAttribute)
+            {
+                CreateEmbeddedAttributeIfNeeded(
+                    builder,
+                    diagnostics,
+                    WellKnownType.Microsoft_CodeAnalysis_EmbeddedAttribute,
+                    WellKnownMember.Microsoft_CodeAnalysis_EmbeddedAttribute__ctor,
+                    allowUserDefined: false);
+
+                CreateEmbeddedAttributeIfNeeded(
+                    builder,
+                    diagnostics,
+                    WellKnownType.System_Runtime_CompilerServices_IsReadOnlyAttribute,
+                    WellKnownMember.System_Runtime_CompilerServices_IsReadOnlyAttribute__ctor,
+                    allowUserDefined: true);
+            }
+
+            return ImmutableDictionary.CreateRange(builder.ToArrayAndFree());
+        }
+
+        private void CreateEmbeddedAttributeIfNeeded(ArrayBuilder<KeyValuePair<WellKnownMember, MethodSymbol>> builder, DiagnosticBag diagnostics, WellKnownType type, WellKnownMember constructor, bool allowUserDefined)
+        {
+            var userDefinedAttribute = _sourceAssembly.DeclaringCompilation.GetWellKnownType(type);
```

Comment:
This is called for other attributes like Embedded as well (and others later), not just IsReadOnly.

---
In reply to: [113573065](https://github.com/dotnet/roslyn/pull/18715#discussion_r113573065) [](ancestors = 113573065)